src/vos/vos_dtx.c

```diff
@@ -929,56 +957,55 @@ vos_dtx_register_record(struct umem_instance *umm, umem_off_t record,
 		struct vos_dtx_act_ent	*dae = dth->dth_ent;
 
 		/* Incarnation log entry implies a share */
-		*tx_id = dae->dae_df_off;
+		*tx_id = DAE_LID(dae);
 		if (type == DTX_RT_ILOG)
 			dth->dth_has_ilog = 1;
 	}
 
 	D_DEBUG(DB_IO, "Register DTX record for "DF_DTI
-		": entry %p, type %d, %s ilog entry, rc %d\n",
-		DP_DTI(&dth->dth_xid), dth->dth_ent, type,
-		dth->dth_has_ilog ? "has" : "has not", rc);
+		": lid=%d entry %p, type %d, %s ilog entry, rc %d\n",
+		DP_DTI(&dth->dth_xid),
+		DAE_LID((struct vos_dtx_act_ent *)dth->dth_ent), dth->dth_ent,
+		type, dth->dth_has_ilog ? "has" : "has not", rc);
 
 	return rc;
 }
 
 /* The caller has started PMDK transaction. */
 void
 vos_dtx_deregister_record(struct umem_instance *umm, daos_handle_t coh,
-			  umem_off_t entry, umem_off_t record)
+			  uint16_t entry, daos_epoch_t epoch, umem_off_t record)
 {
 	struct vos_container		*cont;
 	struct vos_dtx_act_ent		*dae;
 	struct vos_dtx_act_ent_df	*dae_df;
 	struct vos_dtx_record_df	*rec_df;
-	d_iov_t				 kiov;
-	d_iov_t				 riov;
+	bool				 found;
 	int				 count;
-	int				 rc;
 	int				 i;
 
-	if (!vos_dtx_is_normal_entry(umm, entry))
+	if (!vos_dtx_is_normal_entry(entry))
 		return;
 
-	dae_df = umem_off2ptr(umm, entry);
-	if (daos_is_zero_dti(&dae_df->dae_xid) ||
-	    dae_df->dae_flags & DTX_EF_INVALID)
-		return;
+	D_ASSERT(entry >= 2);
 
 	cont = vos_hdl2cont(coh);
 	if (cont == NULL)
-		goto handle_df;
+		return;
```

Comment:
Why "return" instead of original "goto handle_df"? In fact, here, what we worry about is the VOS aggravation. Normally, vos aggregation will not handle non-committed DTX because of there is some time interval between current DTX and VOS aggregation cursor. But if some DTX is blocked for some reason, for example, some replica is too slow or lost respond, then the VOS aggregation may moves its cursor to such DTX. Under such case, dtx_deregister() will be called with NULL 'container'.

Reply:
Ok, will have a look at this case.   Vos aggregation presently treats uncommitted DTX as aborted